sdk/core/src/main/java/com/klaviyo/coresdk/networking/requests/KlaviyoApiRequest.kt

```diff
@@ -0,0 +1,234 @@
+        private const val METHOD_JSON_KEY = "method"
+        private const val TIME_JSON_KEY = "time"
+        private const val UUID_JSON_KEY = "uuid"
+        private const val HEADERS_JSON_KEY = "headers"
+        private const val QUERY_JSON_KEY = "query"
+        private const val BODY_JSON_KEY = "body"
+
+        /**
+         * Construct a request from a JSON object
+         *
+         * Returns [KlaviyoApiRequest] no matter what subclass it was before encoding.
+         * This is functionally the same and saves us some trouble
+         *
+         * @param json
+         * @return
+         */
+        fun fromJson(json: JSONObject): KlaviyoApiRequest {
+            val urlPath = json.getString(PATH_JSON_KEY)
+            val method = when (json.getString(METHOD_JSON_KEY)) {
+                RequestMethod.POST.name -> RequestMethod.POST
+                else -> RequestMethod.GET
+            }
+            val time = json.getString(TIME_JSON_KEY)
+            val uuid = json.getString(UUID_JSON_KEY)
+
+            return KlaviyoApiRequest(urlPath, method, time, uuid).apply {
+                headers = json.getJSONObject(HEADERS_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                query = json.getJSONObject(QUERY_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                body = json.optJSONObject(BODY_JSON_KEY)
+            }
+        }
+
+        /**
+         * Compare two requests purely by their UUIDs
+         *
+         * @param request1
+         * @param request2
+         * @return
+         */
+        fun equals(request1: KlaviyoApiRequest, request2: KlaviyoApiRequest): Boolean =
+            request1.uuid == request2.uuid
+    }
+
+    /**
+     * Compiles the base url, path and query data into a [URL] object
+     */
+    val url: URL
+        get() {
+            val baseUrl = Registry.config.baseUrl
+            val queryMap = query.map { (key, value) -> "$key=$value" }
+            val queryString = queryMap.joinToString(separator = "&")
+
+            return if (queryString.isEmpty())
+                URL("$baseUrl/$urlPath")
+            else
+                URL("$baseUrl/$urlPath?$queryString")
+        }
+
+    /**
+     * Builds and sends a network request to Klaviyo and then parses and handles the response
+     *
+     * @returns The string value of the response body, if one was returned
+     */
+    fun send(): String? {
+        if (!Registry.networkMonitor.isNetworkConnected()) {
+            return null
+        }
+
+        val connection = buildUrlConnection()
+
+        return try {
+            connection.connect()
+            return readResponse(connection)
+        } catch (ex: IOException) {
+            // TODO Logging - TP-164154
+            return null
+        } finally {
+            connection.disconnect()
```

Comment:
can an exception be thrown during `connection.connect()` and will that make disconnect throw an exception too?

Reply:
Great question, I'm going to put a TODO in, I wanted to test the exception handling with the tester app too.

Reply:
Most samples I see just put this in a finally... but I'm paranoid enough that I'm going to put an extra try/catch around `.disconnect()`